File: java/Components/Strata.Domain.Core/src/strata/domain/core/unitofwork/ActiveUnitOfWorkState.java
// ##########################################################################
// # File Name:	ActiveUnitOfWorkState.java
// #
// # Copyright:	2017, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataPersistence Framework.
// #
// #   			The StrataPersistence Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataPersistence Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataPersistence
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.domain.core.unitofwork;

import strata.domain.core.namedquery.INamedQuery;

import java.io.Serializable;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class ActiveUnitOfWorkState
    extends AbstractUnitOfWorkState
{
    private static final ActiveUnitOfWorkState theirInstance = 
        new ActiveUnitOfWorkState();
    
    /************************************************************************
     * Creates a new {@code ActiveUnitOfWorkState}. 
     *
     */
    private 
    ActiveUnitOfWorkState()
    {
        super( "Active" );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<E>
    insertNew(
        AbstractUnitOfWork context,
        Class<K> keyType,
        Class<E> entityType,
        E                  newEntity)
    {
        return context.doInsertNew( keyType,entityType,newEntity );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<E>
    updateExisting(
        AbstractUnitOfWork context,
        Class<K> keyType,
        Class<E> entityType,
        E                  existingEntity)
    {
        return context.doUpdateExisting( keyType,entityType,existingEntity );
    }

    /************************************************************************
     * {@inheritDoc}
     * @return
     */
    @Override
    public <K extends Serializable,E> CompletionStage<E>
    removeExisting(
        AbstractUnitOfWork context,
        Class<K> keyType,
        Class<E> entityType,
        E                  existingEntity)
    {
        return context.doRemoveExisting( keyType,entityType,existingEntity );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Optional<E>>
    getExisting(AbstractUnitOfWork context,Class<E> type,K key)
    {
        return context.doGetExisting( type,key );
    }

    /************************************************************************
     * {@inheritDoc}
     * @return
     */
    @Override
    public <E> CompletionStage<INamedQuery<E>>
    getNamedQuery(
        AbstractUnitOfWork context,
        Class<E> type,
        String queryName)
    {
        return context.doGetNamedQuery( type,queryName );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Boolean>
    hasExisting(AbstractUnitOfWork context,Class<E> type,K key)
    {
        return context.doHasExisting( type,key );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <E> CompletionStage<Boolean>
    hasNamedQuery(
        AbstractUnitOfWork context,
        Class<E> type,
        String queryName)
    {
        return context.doHasNamedQuery( type,queryName );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isActive(AbstractUnitOfWork context)
    {
        return true;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public CompletionStage<Void>
    commit(AbstractUnitOfWork context) 
    {
        return
            context
                .doCommit()
                .thenApply(
                    v ->
                    {
                        context.changeState(CommittedUnitOfWorkState.getInstance());
                        return (Void)null;
                    })
                .exceptionally(
                    e ->
                    {
                        context.changeState( FailedUnitOfWorkState.getInstance() );
                        return null;
                    });
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public CompletionStage<Void>
    rollback(AbstractUnitOfWork context)
    {
        return
            context
                .doRollback()
                .thenApply(
                    v ->
                    {
                        context.changeState(
                            RolledBackUnitOfWorkState.getInstance());
                        return (Void)null;
                    })
                .exceptionally(
                    e ->
                    {
                        context.changeState(FailedUnitOfWorkState.getInstance());
                        return null;
                    });
    }
    
    /************************************************************************
     *  
     *
     * @return
     */
    public static ActiveUnitOfWorkState
    getInstance()
    {
        return theirInstance;
    }
}

// ##########################################################################

File: csharp/Components/Strata.Domain.Core/UnitOfWork/ActiveUnitOfWorkState.cs
//  ##########################################################################
//  # File Name: ActiveUnitOfWorkState.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Strata.Domain.Core.NamedQuery;
using Strata.Foundation.Core.Utility;

namespace Strata.Domain.Core.UnitOfWork
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public
    class ActiveUnitOfWorkState:
        AbstractUnitOfWorkState
    {
        private static readonly ActiveUnitOfWorkState instance = 
            new ActiveUnitOfWorkState();

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>ActiveUnitOfWorkState</c> instance.
        /// </summary>
        /// 
        private
        ActiveUnitOfWorkState():
            base("Active") {}

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void 
        Dispose()
        {
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<E> 
        Insert<K,E>(AbstractUnitOfWork context,E entity)
        {
            return context.DoInsert<K,E>(entity);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<E> 
        Update<K,E>(AbstractUnitOfWork context,E entity)
        {
            return context.DoUpdate<K,E>(entity);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task
        Remove<K,E>(AbstractUnitOfWork context,E entity)
        {
            return context.DoRemove<K,E>(entity);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<Optional<T>> GetUnique<K,T>(AbstractUnitOfWork context,K key)
        {
            return context.DoGetUnique<K,T>( key );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<Optional<T>> GetUniqueMatching<T>(AbstractUnitOfWork context,
            Expression<Func<T,bool>> predicate)
        {
            return context.DoGetUniqueMatching<T>( predicate );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<IList<T>> GetMatching<T>(AbstractUnitOfWork context,
            Expression<Func<T,bool>> predicate)
        {
            return context.DoGetMatching<T>( predicate );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<IList<T>> GetAll<T>(AbstractUnitOfWork context)
        {
            return context.DoGetAll<T>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<INamedQuery<T>> GetNamedQuery<T>(AbstractUnitOfWork context,string queryName)
        {
            return context.DoGetNamedQuery<T>( queryName );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<bool> HasUnique<K,T>(AbstractUnitOfWork context,K key)
        {
            return context.DoHasUnique<K,T>( key );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<bool> HasUniqueMatching<T>(AbstractUnitOfWork context,
            Expression<Func<T,bool>> predicate)
        {
            return context.DoHasMatching<T>( predicate );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override Task<bool> HasNamedQuery<T>(AbstractUnitOfWork context,string queryName)
        {
            return context.DoHasNamedQuery<T>( queryName );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool
        IsActive(AbstractUnitOfWork context)
        {
            return true;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override async Task
        Commit(AbstractUnitOfWork context)
        {
            try
            {
                await context.DoCommit();
                context.State = CommittedUnitOfWorkState.GetInstance();
            }
            catch (Exception e)
            {
                context.State = FailedUnitOfWorkState.GetInstance();
                throw new CommitFailedException( e.Message,e );
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override async Task
        Rollback(AbstractUnitOfWork context)
        {
            try
            {
                await context.DoRollback();
                context.State = RolledBackUnitOfWorkState.GetInstance();
            }
            catch (Exception e)
            {
                throw new RollbackFailedException(e.Message,e);
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// 
        /// </summary>
        /// 
        public static ActiveUnitOfWorkState
        GetInstance()
        {
            return instance;
        }
    }
}

//  ##########################################################################
